pkg/execution/executor/executor.go

```diff
@@ -2556,6 +2558,91 @@ func (e *executor) handleGeneratorAIGateway(ctx context.Context, i *runInstance,
 	return err
 }
 
+// TODO Idempotency
+func (e *executor) handleGeneratorWaitForSignal(ctx context.Context, i *runInstance, gen state.GeneratorOpcode, edge queue.PayloadEdge) error {
+	opts, err := gen.SignalOpts()
+	if err != nil {
+		return fmt.Errorf("unable to parse signal opts: %w", err)
+	}
+	if opts.Signal == "" {
+		return fmt.Errorf("signal name is empty")
+	}
+	expires, err := opts.Expires()
+	if err != nil {
+		return fmt.Errorf("unable to parse signal expires: %w", err)
+	}
+
+	pauseID := inngest.DeterministicSha1UUID(i.md.ID.RunID.String() + gen.ID)
+	opcode := gen.Op.String()
+	now := time.Now()
+
+	sid := run.NewSpanID(ctx)
+	carrier := itrace.NewTraceCarrier(
+		itrace.WithTraceCarrierTimestamp(now),
+		itrace.WithTraceCarrierSpanID(&sid),
+	)
+	itrace.UserTracer().Propagator().Inject(ctx, propagation.MapCarrier(carrier.Context))
+
+	_, err = e.pm.SavePause(ctx, state.Pause{
+		ID:          pauseID,
+		WorkspaceID: i.md.ID.Tenant.EnvID,
+		Identifier:  sv2.NewPauseIdentifier(i.md.ID),
+		GroupID:     i.item.GroupID,
+		Outgoing:    gen.ID,
+		Incoming:    edge.Edge.Incoming,
+		StepName:    gen.UserDefinedName(),
+		Opcode:      &opcode,
+		Expires:     state.Time(expires),
+		DataKey:     gen.ID,
+		SignalID:    &opts.Signal,
+		MaxAttempts: i.item.MaxAttempts,
+		Metadata: map[string]any{
+			consts.OtelPropagationKey: carrier,
+		},
+	})
+	if err == state.ErrPauseAlreadyExists {
+		return nil
+	}
+	if err == state.ErrSignalConflict {
```

Comment:
In what case can we have `state.ErrSignalConflict` and `state.ErrPauseAlreadyExists` if saving signals is idempotent?

Reply:
`ErrPauseAlreadyExists` means this was an idempotent retry and we're all set - no further action needed.

`ErrSignalConflict` means that a wait for that signal already exists but for _another_ run, indicating that we must fail this particular run with a non-retriable error.

In fact, that's probably a point - this may not be non-retriable. I'll make sure.